doc/development/plugins.rst

```diff
@@ -97,169 +92,165 @@ as well as potential further capabilities, by providing the following class attr
-the :meth:`~.Device.expval` method will be called as follows from PennyLane:
+Once all the class attributes are defined, it is necessary to define some required class
+methods, to allow PennyLane to apply operations and measure observables on your device.
 
-.. code-block:: python
+To execute operations on the device, the following methods **must** be defined:
 
-    result = dev.expval(["PauliX, Hermitian"], wires=[[0], [1, 3]], par=[[], [A]])
+.. currentmodule:: pennylane.QubitDevice
 
+.. autosummary::
 
-Advanced execution control
-^^^^^^^^^^^^^^^^^^^^^^^^^^
+    apply
+    probability
 
-Additional flexibility is sometimes required for interfacing with more
-complicated frameworks.
+(Click the name of each method to see more details and examples). The :class:`~.QubitDevice` class
+provides the methods :meth:`~.QubitDevice.active_wires` and
+:meth:`~.QubitDevice.marginal_prob` for convenience.
 
-When PennyLane needs to evaluate a QNode, it accesses the :meth:`~.Device.execute` method of
-your plugin, which, by default performs the following process:
+In addition, if your device generates its own computational basis samples for measured modes
+after execution, you need to overwrite the following method:
 
-.. code-block:: python
+.. autosummary::
+
+    generate_samples
 
-    results = []
+Note that ``self._samples`` has the shape ``(dev.shots, dev.num_wires)``.
+Furthermore, PennyLane uses the convention :math:`|q_0,q_1,\dots,q_{N-1}\rangle` where
+:math:`q_0` is the most significant bit.
 
-    with self.execution_context():
-        self.pre_apply()
-        for operation in queue:
-            self.apply(operation.name, operation.wires, operation.parameters)
-        self.post_apply()
+And thats it! The device has inherited :meth:`~.QubitDevice.expval`, :meth:`~.QubitDevice.var`,
+and :meth:`~.QubitDevice.sample` methods, that accepts an observable (or tensor product of
+observables) and returns the corresponding measurement statistic.
 
-        self.pre_measure()
 
-        for obs in observables:
-            if obs.return_type is Expectation:
-                results.append(self.expval(obs.name, obs.wires, obs.parameters))
 
-            elif obs.return_type is Variance:
-                results.append(self.var(obs.name, obs.wires, obs.parameters))
+:html:`<div class="caution admonition" id="aside1"><a data-toggle="collapse" data-parent="#aside1" href="#content1" class="collapsed"><p class="first admonition-title">Advanced execution control (click to expand) <i class="fas fa-chevron-circle-down"></i></p></a><div id="content1" class="collapse" data-parent="#aside1" style="height: 0px;">`
 
-            elif obs.return_type is Sample:
-                results.append(np.array(self.sample(obs.name, obs.wires, obs.parameters)))
+Additional flexibility is sometimes required for interfacing with more
+complicated frameworks.
+
+When PennyLane needs to evaluate a QNode, it accesses the :meth:`~.QubitDevice.execute` method of
```

Comment:
```suggestion
When PennyLane needs to evaluate a :class:`~.QNode`, it accesses the :meth:`~.QubitDevice.execute` method of
```

Reply:
(Perhaps it was not referenced since it is a wrapper around the ``QNode`` classes?)

Reply:
Exactly, since the QNode refactor there isn't one specific object that makes sense to reference here any more 😆